sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +289,236 @@ def reduce(self, word, exclude=None):
+
+        # Create the states in the automaton.
+        # The left-hand side of the rules are the dead states.
+        # The proper left-hand side of the rules are the accept states.
+
+        # Create a finite state machine as an instance of the StateMachine object
+        fsm = StateMachine('fsm')
+        fsm.add_state('start', is_start=True)
+
+        # Add dead states.
+        for rule in left_hand_rules:
+            if not rule in fsm.states:
+                fsm.add_state(rule, is_dead=True)
+
+        # Add accept states.
+        for r in all_rules:
+            prop_prefix = proper_prefixes[r]
+            for elem in prop_prefix:
+                if not elem in fsm.states:
+                    fsm.add_state(elem, is_accept=True)
+
+        # Add transitions for every state
+        for state in fsm.states:
+            current_state_name = state
+            current_state_type = fsm.states[state].state_type
+            if current_state_type == "start":
+                for letter in automaton_alphabet:
+                    if letter in fsm.states:
+                        fsm.states[state].add_transition(letter, letter)
+                    else:
+                        fsm.states[state].add_transition(letter, current_state_name)
+            elif current_state_type == "accept":
+                for letter in automaton_alphabet:
+                    next = current_state_name*letter
+                    len_next_word = len(next)
+                    while True:
+                        if len(next) <= 1:
+                            if next in fsm.states:
+                                fsm.states[state].add_transition(letter, next)
+                            else:
+                                fsm.states[state].add_transition(letter, 'start')
+                            break
+                        else:
+                            if next in fsm.states:
+                                fsm.states[state].add_transition(letter, next)
+                                break
+                            next = next.subword(1, len(next))
+        return fsm
+
+    def add_to_automaton(self, rules):
+        # Automaton vairables
+        automaton_alphabet = []
+        left_hand_rules = []
+        proper_prefixes = {}
+
+        # compute the inverses of all the new rules added
+        all_rules = rules
+        inverse_rules = self.compute_inverse_rules(all_rules)
+        all_rules.update(inverse_rules)
+
+        # All the transition symbols in the automaton
+        generators = list(self.alphabet)
+        generators += [gen**-1 for gen in generators]
+
+        # The symbols present in the new rules are the symbols to be verified at each state.
+        # computes the automaton_alphabet, as the transitions solely depend upon the new states.
+        for rule in rules:
+            automaton_alphabet += rule.letter_form_elm
+        automaton_alphabet = set(automaton_alphabet)
```

Comment:
I think you should make `FiniteStateMachine` have an `alphabet` attribute, then this will just be `self.reduction_automaton.alphabet` (you already compute it in other places so seems inefficient to do it again, and it makes sense for a fsm to have its alphabet as an attribute).

Reply:
Makes sense. But in this case, the transitions of the old states solely depend on the symbols of the new state. For example, if the `automaton_alphabet`is `[a, a**-1, b, b**-1]` and the newly added rule only consists of the symbol `a`, the `automaton_alphabet` should be recomputed, so that the transitions(at the previously added old states) are only verified for the symbols in the newly added rules.

Reply:
I think that `automaton_alphabet` variable has to be changed in the `add_to_automaton` function. That could be misleading as it doesn't store all the symbols in the `automaton_alphabet`.

Reply:
Ah, I see. In that case what you should do here depends on what you think about the thing I suggested for adding new transitions later. If you decide to construct a dictionary similar to what I described, then you could extract the letters you want from there.